client/src/components/Game/GameProvider.tsx

```diff
@@ -0,0 +1,176 @@
+import useSocketClient from "hooks/useSocketClient";
+import IPlayerConnected from "components/models/websocket/IPlayerConnected";
+import { Game as GameInstance } from "chess-easy";
+import IChatMessage from "components/models/websocket/IChatMessage";
+
+interface GameProviderContext {
+  game: IGame;
+  color: string;
+  players: IPlayerConnected;
+  chat: IChatMessage[];
+  setChat: React.Dispatch<React.SetStateAction<IChatMessage[]>>;
+  sendMove: (from: string, to: string, promotion?: string) => Promise<boolean>;
+  gameInstance: GameInstance;
+}
+
+interface GameContextProviderProps {
+  children: React.ReactNode;
+}
+
+interface Move {
+  from: string;
+  to: string;
+  promotion: string;
+}
+
+const GameContext = createContext<GameProviderContext | null>(null);
+
+export const GameContextProvider = ({ children }: GameContextProviderProps) => {
+  const [game, setGame] = useState<IGame>();
+  const [color, setColor] = useState<string>();
+  const [players, setPlayers] = useState<IPlayerConnected>();
+  const axiosPrivate = useAxiosPrivate();
+  const navigate = useNavigate();
+  const auth = useAuth().auth;
+  const gameId = useParams().id;
+  const socket = useSocketClient();
+  const [isLoading, setIsLoading] = useState(false);
+  const [chat, setChat] = useState<IChatMessage[]>([]);
+  const [gameInstance, setGameInstance] = useState<GameInstance>();
+  const [lastMove, setLastMove] = useState<Move>();
+
+  useLayoutEffect(() => {
+    const joinGame = async () => {
+      try {
+        setIsLoading(true);
+        const joinAttempt = await axiosPrivate.put(`/games/join/${gameId}`, {
+          auth,
+        });
+
+        if (joinAttempt.status === 200) {
+          const { fen, chat, playerWhite, playerBlack, _id, ...rest } =
+            joinAttempt.data.game;
+          setChat(chat);
+          setPlayers({ playerBlack, playerWhite });
+          const gameInstance = new GameInstance(...(fen ? [fen] : []));
+          setGameInstance(gameInstance);
+          setGame({
+            ...rest,
+            id: _id,
+            gameState: gameInstance.getGameStateObject(),
+            possibleMoves: getPossibleMoves(gameInstance),
+          });
+          if (playerWhite?.username === auth.username) {
+            setColor("white");
+          } else if (playerBlack?.username === auth.username) {
+            setColor("black");
+          }
+        }
+        setIsLoading(false);
+      } catch (error) {
+        console.error(error);
+        navigate("/lobby");
+      }
+    };
+
+    joinGame();
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const getPossibleMoves = useCallback(
+    (gameInstance: GameInstance) => {
+      if (!players || !Object.values(players).every(value => value))
+        return null;
+      return gameInstance.getNextColor() === color
+        ? gameInstance.possibleMoves
+        : null;
+    },
+    [color, players],
+  );
```

Comment:
Czemu useCallback tutaj?

Reply:
To w useEffecie wykorzystujemy, to musi byc useCallback, bo bez tego inifnite loop idzie.